### PR TITLE
Generate: contrastive search uses existing abstractions and conventions

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?
This PR updates contrastive search to follow existing abstractions conventions in other generation functions. It consists of several tiny changes, with the reasoning for each change in the PR comments below.

This is part of the effort to make converting to TF easier. All slow tests pass (`RUN_SLOW=1 py.test tests/generation/test_generation_utils.py -k contrastive -vv`)

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -1882,28 +1882,34 @@ def contrastive_search(
...
             # if the first step in the loop, encode all the prefix and obtain three parameters: (1) past_key_values;
-            # (2) last_hidden_states; (3) logit_for_next_step
+            # (2) last_hidden_states; (3) logit_for_next_step; (4) update model kwargs for the next step
             if model_kwargs.get("past") is None:
                 # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save
                 # the `encoder_outputs`
-                output = self(**model_inputs, output_hidden_states=True, output_attentions=True)
+                outputs = self(
+                    **model_inputs, return_dict=True, output_hidden_states=True, output_attentions=output_attentions
+                )
```
Comment (gante) [inline src/transformers/generation_utils.py:1891]: Two changes here:
1. makes `return_dict` explicitly true (needed for contrastive search) and `output_attentions` optional (as in the other functions)
2. `output` -> `outputs`, as in the other functions (to facilitate copy/paste)

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -1882,28 +1882,34 @@ def contrastive_search(
...
                 else:
-                    last_hidden_states = output.hidden_states[-1]
+                    last_hidden_states = outputs.hidden_states[-1]
                 # next logit for contrastive search to select top-k candidate tokens
-                logit_for_next_step = output.logits[:, -1, :]
+                logit_for_next_step = outputs.logits[:, -1, :]
+
+                model_kwargs = self._update_model_kwargs_for_generation(
+                    outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
+                )
```
Comment (gante) [inline src/transformers/generation_utils.py:1912]: The most important change: updating the attention mask was happening manually below, and we already had a function that took care of it. While updating the mask is simple in PT, in TF with XLA this amounts to >100 lines of code.

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -1918,6 +1924,18 @@ def contrastive_search(
...
+            if return_dict_in_generate:
+                if output_scores:
+                    scores += (logit_for_next_step,)
+
+                if output_hidden_states:
+                    decoder_hidden_states += (
+                        (outputs.decoder_hidden_states,)
+                        if self.config.is_encoder_decoder
+                        else (outputs.hidden_states,)
+                    )
```
Comment (gante) [inline src/transformers/generation_utils.py:1937]: Three changes:
1. Moves the `return_dict_in_generate` block to where it happens in the other generate functions (after getting the logits)
2. Stores the correct hidden states: before it was storing the hidden states for generation step X+1, as opposed to the intended step X
3. Stores the correct scores: I incorrectly suggested to the authors to store the post-contrastive search penalty scores here. Those scores are definitely not logits, so it would break the API with respect to other generation methods. Instead, store the logits here, as in the other functions. (note: we may decide to include the post-contrastive search penalty scores in the future, as part of the output)

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -1958,27 +1973,26 @@ def contrastive_search(
...
                 last_hidden_states.unsqueeze(1).expand(-1, top_k, -1, -1).reshape(bsz * top_k, seqlen, embed_dim)
             )
 
             # compute the degeneratin penalty and re-rank the candidates based on the degeneration penalty and the
             # model confidence
-            # the scores and index of the selected tokens are returned
-            selected_scores, selected_idx = ranking_fast(
-                context_hidden, next_hidden, top_k_probs, penalty_alpha, top_k
-            )
+            selected_idx = _ranking_fast(context_hidden, next_hidden, top_k_probs, penalty_alpha, top_k)
```
Comment (gante) [inline src/transformers/generation_utils.py:1995]: As written above, post-contrastive search penalty scores are no longer needed.

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -2009,21 +2023,24 @@ def contrastive_search(
...
+                outputs = Seq2SeqLMOutput(
+                    past_key_values=past_key_values,
+                    decoder_hidden_states=decoder_hidden_states,
+                )
+            else:
+                outputs = CausalLMOutputWithPast(
+                    past_key_values=past_key_values,
+                    hidden_states=decoder_hidden_states,
+                    attentions=model_kwargs["attention_mask"],
+                )
```
Comment (gante) [inline src/transformers/generation_utils.py:2038]: moves this above the "contrastive_search main logic end" comment, since it is unique to contrastive search

**Review by sgugger — APPROVED** [review]
Thanks for updating!
